libsolidity/experimental/ast/TypeSystem.cpp

```diff
@@ -0,0 +1,351 @@
+			if (_left.index() == _right.index())
+			{
+				solAssert(_left.sort() == _right.sort());
+				return true;
+			}
+			return false;
+		},
+		[&](TypeConstant _left, TypeConstant _right) {
+			if (_left.constructor != _right.constructor)
+				return false;
+			if (_left.arguments.size() != _right.arguments.size())
+				return false;
+			for (auto&& [left, right]: ranges::zip_view(_left.arguments, _right.arguments))
+				if (!typeEquals(left, right))
+					return false;
+			return true;
+		},
+		[&](auto, auto) {
+			return false;
+		}
+	}, resolve(_lhs), resolve(_rhs));
+}
+
+TypeEnvironment TypeEnvironment::clone() const
+{
+	TypeEnvironment result{m_typeSystem};
+	result.m_typeVariables = m_typeVariables;
+	return result;
+}
+
+TypeSystem::TypeSystem()
+{
+	auto declarePrimitiveClass = [&](std::string _name) {
+		return std::visit(util::GenericVisitor{
+			[](std::string _error) -> TypeClass {
+				solAssert(false, _error);
+			},
+			[](TypeClass _class) -> TypeClass { return _class; }
+		}, declareTypeClass(freshVariable({}), _name, nullptr));
+	};
+
+	m_primitiveTypeClasses.emplace(PrimitiveClass::Type, declarePrimitiveClass("type"));
+
+	for (auto [type, name, arity]: std::initializer_list<std::tuple<PrimitiveType, char const*, size_t>>{
+		{PrimitiveType::TypeFunction, "tfun", 2},
+		{PrimitiveType::Function, "fun", 2},
+		{PrimitiveType::Itself, "itself", 1},
+		{PrimitiveType::Void, "void", 0},
+		{PrimitiveType::Unit, "unit", 0},
+		{PrimitiveType::Pair, "pair", 2},
+		{PrimitiveType::Sum, "sum", 2},
+		{PrimitiveType::Word, "word", 0},
+		{PrimitiveType::Integer, "integer", 0},
+		{PrimitiveType::Bool, "bool", 0},
+	})
+		m_primitiveTypeConstructors.emplace(type, declareTypeConstructor(name, name, arity, nullptr));
+
+	TypeClass classType = primitiveClass(PrimitiveClass::Type);
+	//TypeClass classKind = primitiveClass(PrimitiveClass::Kind);
+	Sort typeSort{{classType}};
+	m_typeConstructors.at(m_primitiveTypeConstructors.at(PrimitiveType::TypeFunction).m_index).arities = {Arity{std::vector<Sort>{{typeSort},{typeSort}}, classType}};
+	m_typeConstructors.at(m_primitiveTypeConstructors.at(PrimitiveType::Function).m_index).arities = {Arity{std::vector<Sort>{{typeSort, typeSort}}, classType}};
+	m_typeConstructors.at(m_primitiveTypeConstructors.at(PrimitiveType::Function).m_index).arities = {Arity{std::vector<Sort>{{typeSort, typeSort}}, classType}};
+}
+
+experimental::Type TypeSystem::freshVariable(Sort _sort)
+{
+	size_t index = m_numTypeVariables++;
+	return TypeVariable(index, std::move(_sort));
+}
+
+experimental::Type TypeSystem::freshTypeVariable(Sort _sort)
+{
+	_sort.classes.emplace(primitiveClass(PrimitiveClass::Type));
+	return freshVariable(_sort);
+}
+
+std::vector<TypeEnvironment::UnificationFailure> TypeEnvironment::instantiate(TypeVariable _variable, Type _type)
+{
+	for (auto typeVar: TypeEnvironmentHelpers{*this}.typeVars(_type))
+		if (typeVar.index() == _variable.index())
```

Comment:
This looks like a bug. `typeVars()` returns a vector of `Type` rather than `TypeVariable` so here you're calling [`std::variant::index()`](https://en.cppreference.com/w/cpp/utility/variant/index) rather than [`TypeVariable::index()`].

Perhaps we should rename `index()` to `id()` to avoid this kind of mistake in the future?

Reply:
And by the way, for some reason you often return `Type` when what you have is definitely a type variable. For example `freshTypeVariable()` returns `Type`. I'm going to change that because it forces you to bother with unpacking the variant when you want to actually use it as a variable, which is completely unnecessary.

Same here, `typeVars()` looks like it should really be returning `vector<TypeVariable>`. And it would have prevented this bug too.

Reply:
This is now a part of #14658.